### DISCUSSION TITLE
User Interface with a window manager as in Transport Tycoon

### DISCUSSION THREAD

**Discussion by vault-thirteen** [MonoGame/MonoGame#8604] [General]
Hello there.

I am searching for new ways to create a simple 2D computer game with a rich user interface and internet users tell me that MonoGame is quite good for 2D games. 

I want to create a 2D game with a GUI (graphical user interface) similar to the Transport Tycoon (a.k.a. OpenTTD) game. Does MonoGame provide tools for that ? If not, can you direct me to good libraries which are compatible with MonoGame ?

If you do not know what is the GUI in Transport Tycoon, I will show a screenshot and describe what I need from it.

![1](https://github.com/user-attachments/assets/0c0b2a82-83d1-46b2-878b-6ed993c33083)

Here we can see that there are a lot of windows in the game which allow us to control a lot of parameters of numerous objects. So, the game uses its own window manager (W.M.). This W.M. supports many useful features. I will try to list some of the most useful features of it.

1. Windows can snap to each other.
2. Windows can be resized with a special button in the bottom right corner.
3. Windows can be moved by their header.
4. Windows can be closed with the [x] button.
5. Windows can contain elements such as buttons, labels, text, lists with scrollbar, tables and many more.
6. The most unique feature of this game is that its W.M. allows to draw game map inside the windows, so we can see where our vehicle is driving at the moment. So, windows support graphical content inside of them which is taken from some buffer or map.

I am searching for a way to implement a similar window manager in my 2D game using C# and MonoGame. Is it possible ? Are there any helper libraries for this ? Or should I re-invent the wheel again ?

Thank you.

**Comment by squarebananas** [comment]
Here are some UI libraries for MonoGame: https://github.com/Apostolique/MonoGameLibraries/?tab=readme-ov-file#user-interface

Also it seems that [game's UI is open source](https://github.com/OpenTTD/OpenTTD/blob/master/src/widget.cpp), however it's written in C++ so would need a fair amount of work to port to MonoGame.

**Reply by MonogameGuy** [reply]
Be careful with the GPL2 licensing on the source code to "OpenTTD" if you do not plan on releasing an open source game.
